authentication/src/main/java/io/camunda/authentication/CamundaOidcUserService.java

```diff
@@ -0,0 +1,81 @@
+import io.camunda.search.entities.MappingEntity;
+import io.camunda.service.AuthorizationServices;
+import io.camunda.service.GroupServices;
+import io.camunda.service.MappingServices;
+import io.camunda.service.RoleServices;
+import io.camunda.service.TenantServices;
+import io.camunda.service.TenantServices.TenantDTO;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.Profile;
+import org.springframework.security.oauth2.client.oidc.userinfo.OidcUserRequest;
+import org.springframework.security.oauth2.client.oidc.userinfo.OidcUserService;
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.core.oidc.user.OidcUser;
+import org.springframework.stereotype.Service;
+
+@Service
+@Profile("auth-oidc")
+public class CamundaOidcUserService extends OidcUserService {
+  private static final Logger LOG = LoggerFactory.getLogger(CamundaOidcUserService.class);
+  private final MappingServices mappingServices;
+  private final TenantServices tenantServices;
+  private final RoleServices roleServices;
+  private final GroupServices groupServices;
+  private final AuthorizationServices authorizationServices;
+
+  public CamundaOidcUserService(
+      final MappingServices mappingServices,
+      final TenantServices tenantServices,
+      final RoleServices roleServices,
+      final GroupServices groupServices,
+      final AuthorizationServices authorizationServices) {
+    this.mappingServices = mappingServices;
+    this.tenantServices = tenantServices;
+    this.roleServices = roleServices;
+    this.groupServices = groupServices;
+    this.authorizationServices = authorizationServices;
+  }
+
+  @Override
+  public OidcUser loadUser(final OidcUserRequest userRequest) throws OAuth2AuthenticationException {
+    final OidcUser oidcUser = super.loadUser(userRequest);
+
+    final Map<String, Object> claims = userRequest.getIdToken().getClaims();
+    final List<MappingEntity> mappings = mappingServices.getMatchingMappings(claims);
+    final Set<Long> mappingKeys =
+        mappings.stream().map(MappingEntity::mappingKey).collect(Collectors.toSet());
+    if (mappingKeys.isEmpty()) {
+      LOG.debug("No mappings found for these claims: {}", claims);
+    }
+
+    return new CamundaOidcUser(
+        oidcUser,
+        mappingKeys,
+        new AuthenticationContext(
+            roleServices.getRolesByMemberKeys(mappingKeys),
+            authorizationServices.getAuthorizedApplications(mappingKeys),
+            tenantServices.getTenantsByMemberKeys(mappingKeys).stream()
+                .map(TenantDTO::fromEntity)
+                .toList(),
+            groupServices.getGroupsByMemberKeys(mappingKeys).stream()
+                .map(GroupEntity::name)
+                .toList()));
```

Comment:
💭 Roman raised to me once before that we effectively perform the calls multiple times, once here and then also again down in the engine. I think for now this is fine but we should look to optimize this and only retrieve them whenever we really need to (either by the auth checker, or the engine). Here is the thread https://camunda.slack.com/archives/C06UYJMMETZ/p1734351732320729?thread_ts=1734343187.803619

Reply:
Are there actually many requests that don't need most of the authentication context?

Reply:
Yea exactly, so the decision would be that any call that results in an engine command (create, update, delete) will check the auth context directly in the engine, searches would need the context.

Reply:
Raised https://github.com/camunda/camunda/issues/26739